src/main/java/frc/robot/utilityClasses/CircleFitter.java

```diff
@@ -5,70 +5,81 @@
+        // calculate matrices
+        xsum = ysum = xSquaredsum = ySquaredsum = xysum = index1 = index2 = index3 = 0;
 
-    private static void computeSums() {
-        for (int i = 0; i < circlePoints.size(); i++) {
-            totalX += circlePoints.get(i).getX();
-            totalXSquared += Math.pow(circlePoints.get(i).getY(), 2);
-            totalY += circlePoints.get(i).getY();
-            totalYSquared += Math.pow(circlePoints.get(i).getY(), 2);
-            totalXY += circlePoints.get(i).getX() * circlePoints.get(i).getX();
-            index1 +=
-                    circlePoints.get(i).getX()
-                            * (Math.pow(circlePoints.get(i).getX(), 2)
-                                    + Math.pow(circlePoints.get(i).getY(), 2));
-            index2 +=
-                    circlePoints.get(i).getY()
-                            * (Math.pow(circlePoints.get(i).getX(), 2)
-                                    + Math.pow(circlePoints.get(i).getY(), 2));
-            index3 +=
-                    (Math.pow(circlePoints.get(i).getX(), 2)
-                            + Math.pow(circlePoints.get(i).getY(), 2));
+        for (int i = 0; i < points.size(); i++) {
+            GlobalPoint point = points.get(i);
+            double x = point.x;
+            double y = point.y;
+
+            xsum += x;
+            ysum += y;
+            xSquaredsum += x * x;
+            ySquaredsum += y * y;
+            xysum += x * y;
+
+            index1 += x * (x * x + y * y);
+            index2 += y * (x * x + y * y);
+            index3 += x * x + y * y;
         }
-    }
+        // set matrices
+        X.set(0, 0, xSquaredsum);
+        X.set(0, 1, xysum);
+        X.set(0, 2, xsum);
 
-    private static double[] computeCenter() {
-        a.set(0, 0, totalXSquared);
-        a.set(0, 1, totalXY);
-        a.set(0, 2, totalX);
-        a.set(1, 0, totalXY);
-        a.set(1, 1, totalYSquared);
-        a.set(1, 2, totalY);
-        a.set(2, 0, totalX);
-        a.set(2, 1, totalY);
-        a.set(2, 2, circlePoints.size());
-
-        b.set(0, 0, index1);
-        b.set(1, 0, index2);
-        b.set(2, 0, index3);
-        c = a.invert().mult(b);
-        return new double[2];
+        X.set(1, 0, xysum);
+        X.set(1, 1, ySquaredsum);
+        X.set(1, 2, ysum);
+
+        X.set(2, 0, xsum);
+        X.set(2, 1, ysum);
+        X.set(2, 2, circlePoints.size());
+
+        Y.set(0, 0, index1);
+        Y.set(1, 0, index2);
+        Y.set(2, 0, index3);
     }
 
-    public static void main(String args[]) {
-        computeSums();
-        computeCenter();
+    public static double[] calculateCircle() {
+        double[] ret = new double[3];
+
+        SimpleMatrix M = X.invert().mult(Y);
+
+        A = M.get(0);
+        B = M.get(1);
+        C = M.get(2);
+
+        // k value
+        ret[0] = A / 2;
+        // m value
+        ret[1] = B / 2;
+        // radius
+        ret[2] = Math.sqrt(4 * C + A * A + B * B) / 2;
```

Comment:
We know the radius of the goal (it is in the game manual). Check the radius calculated against our actual one and if it is outside of a certain range, then set radius to 0.

Reply:
I've seen least squares circle estimation with an enforced radius, but they're all in sigma notation and I'm not advanced enough to turn it into matrices. Is this what you mean or we're tossing the data if the radius difference is too large?

Reply:
After some research, I believe that using a fixed radius will bias the center based on the distribution of data points. Since the robot will be looking at the target from the side, we'll be highly susceptible to this. I think your initial suggestion of throwing out the center if the radius is too far off is the best approach